engine/src/main/java/org/hibernate/validator/internal/properties/javabean/JavaBeanExecutable.java

```diff
@@ -183,14 +183,39 @@ private static List<JavaBeanParameter> getParameters(Executable executable) {
 			return Collections.emptyList();
 		}
 
-		List<JavaBeanParameter> parameters = new ArrayList<>( executable.getParameterCount() );
+		int parameterCount = executable.getParameterCount();
+		List<JavaBeanParameter> parameters = new ArrayList<>( parameterCount );
 
 		Parameter[] parameterArray = executable.getParameters();
 		Class<?>[] parameterTypes = executable.getParameterTypes();
 		AnnotatedType[] annotatedTypes = executable.getAnnotatedParameterTypes();
 
+		Annotation[][] parameterAnnotationsArray = executable.getParameterAnnotations();
+		Annotation[][] annotationsForJDK8303112 = null;
+		if ( parameterAnnotationsArray.length != parameterCount ) {
+			// This is a workaround for https://bugs.openjdk.org/browse/JDK-8303112
+			// We're in a situation where parameter.getAnnotation()/parameter.getAnnotations()
+			// is buggy when there are implicit/synthetic parameters,
+			// because constructor.getParameterAnnotations() (wrongly) ignores implicit/synthetic parameters
+			// while parameter.getAnnotations() (rightly) assumes they are present in the array.
+			annotationsForJDK8303112 = new Annotation[parameterCount][];
+			int nonImplicitNorSyntheticParamIndex = 0;
+			for ( int i = 0; i < parameterCount; i++ ) {
+				Parameter parameter = parameterArray[i];
+				if ( parameter.isImplicit() || parameter.isSynthetic() ) {
+					annotationsForJDK8303112[i] = JavaBeanParameter.EMPTY_PARAMETER_ANNOTATIONS;
+				}
+				else {
+					annotationsForJDK8303112[i] =
+							parameterAnnotationsArray[nonImplicitNorSyntheticParamIndex];
+					++nonImplicitNorSyntheticParamIndex;
+				}
+			}
+		}
+
 		for ( int i = 0; i < parameterArray.length; i++ ) {
-			parameters.add( new JavaBeanParameter( i, parameterArray[i], parameterTypes[i], annotatedTypes[i] ) );
+			parameters.add( new JavaBeanParameter( i, parameterArray[i], parameterTypes[i], annotatedTypes[i],
```

Comment:
Maybe it's worth just removing the `Parameter parameter` from `JavaBeanParameter` and always constructing this array of annotations? These annotations are iterated in any case when metadata is created.

Reply:
Maybe... I went for the approach that affects as few applications as possible to avoid introducing regressions, but we can always pile up performance improvements on to of that.